File: internal/spatial/hrir/hrir.go
// Package hrir carries the measured head-related impulse responses occam
// renders with, and nothing else.
//
// Source: the SADIE II database, subject D1, a Neumann KU100 dummy head,
// measured at the University of York. 48 kHz, 256 taps, CC BY 4.0. See
// LICENSE-SADIE.md for the attribution that licence requires.
//
// Only the directions occam's speaker layouts actually use are embedded,
// thirteen of the 8802 measured. Each one landed on an exact measured
// position, so none of them are interpolated.
package hrir

import (
	_ "embed"
	"encoding/binary"
	"fmt"
	"math"
)

//go:embed sadie48.bin
var blob []byte

// Response is one measured direction: the impulse arriving at each ear.
type Response struct {
	Azimuth   float64
	Elevation float64
	Left      []float64
	Right     []float64
}

// Set is everything embedded, at one sample rate.
type Set struct {
	Rate      int
	Taps      int
	Responses []Response
}

const magic = "OCHR"

// Load parses the embedded blob. It is cheap enough to call once at startup
// and there is nothing to fail at runtime, so errors here mean a corrupt build.
func Load() (Set, error) {
	if len(blob) < 20 || string(blob[0:4]) != magic {
		return Set{}, fmt.Errorf("hrir: embedded data is not an %s blob", magic)
	}
	version := binary.LittleEndian.Uint32(blob[4:8])
	if version != 1 {
		return Set{}, fmt.Errorf("hrir: blob version %d, this build reads 1", version)
	}

	set := Set{
		Rate: int(binary.LittleEndian.Uint32(blob[8:12])),
		Taps: int(binary.LittleEndian.Uint32(blob[12:16])),
	}
	count := int(binary.LittleEndian.Uint32(blob[16:20]))

	at := 20
	stride := 8 + set.Taps*4*2
	if want := 20 + count*stride; len(blob) != want {
		return Set{}, fmt.Errorf("hrir: blob is %d bytes, expected %d", len(blob), want)
	}

	set.Responses = make([]Response, count)
	for i := range set.Responses {
		r := Response{
			Azimuth:   float64(math.Float32frombits(binary.LittleEndian.Uint32(blob[at : at+4]))),
			Elevation: float64(math.Float32frombits(binary.LittleEndian.Uint32(blob[at+4 : at+8]))),
			Left:      make([]float64, set.Taps),
			Right:     make([]float64, set.Taps),
		}
		p := at + 8
		for t := range set.Taps {
			r.Left[t] = float64(math.Float32frombits(binary.LittleEndian.Uint32(blob[p : p+4])))
			p += 4
		}
		for t := range set.Taps {
			r.Right[t] = float64(math.Float32frombits(binary.LittleEndian.Uint32(blob[p : p+4])))
			p += 4
		}
		set.Responses[i] = r
		at += stride
	}
	return set, nil
}

// Nearest returns the closest embedded direction and the angle it is off by,
// in degrees. Every direction occam's layouts use is exact, so a non-zero
// error means a layout gained a speaker the blob was not generated for.
func (s Set) Nearest(azimuth, elevation float64) (Response, float64) {
	best, bestDot := Response{}, -2.0
	target := unit(azimuth, elevation)
	for _, r := range s.Responses {
		if d := dot(target, unit(r.Azimuth, r.Elevation)); d > bestDot {
			best, bestDot = r, d
		}
	}
	if bestDot > 1 {
		bestDot = 1
	}
	return best, math.Acos(bestDot) * 180 / math.Pi
}

type vec3 [3]float64

func unit(az, el float64) vec3 {
	a, e := az*math.Pi/180, el*math.Pi/180
	return vec3{math.Cos(e) * math.Cos(a), math.Cos(e) * math.Sin(a), math.Sin(e)}
}

func dot(a, b vec3) float64 { return a[0]*b[0] + a[1]*b[1] + a[2]*b[2] }

File: occam-live/src/hrir.rs
// Shares the blob with the Go renderer, so a difference between the two is
// never the data. SADIE II D1, CC BY 4.0, attribution in
// ../../internal/spatial/hrir/LICENSE-SADIE.md.

const BLOB: &[u8] = include_bytes!("../../internal/spatial/hrir/sadie48.bin");
const MAGIC: &[u8; 4] = b"OCHR";

pub struct Response {
    pub azimuth: f32,
    pub elevation: f32,
    pub left: Vec<f32>,
    pub right: Vec<f32>,
}

pub struct Set {
    pub rate: u32,
    pub taps: usize,
    pub responses: Vec<Response>,
}

impl Set {
    pub fn load() -> Result<Self, String> {
        if BLOB.len() < 20 || &BLOB[0..4] != MAGIC {
            return Err("embedded hrir data is not an OCHR blob".into());
        }
        let u32_at = |at: usize| {
            u32::from_le_bytes([BLOB[at], BLOB[at + 1], BLOB[at + 2], BLOB[at + 3]])
        };
        let f32_at = |at: usize| f32::from_bits(u32_at(at));

        if u32_at(4) != 1 {
            return Err(format!("hrir blob version {}, this build reads 1", u32_at(4)));
        }
        let rate = u32_at(8);
        let taps = u32_at(12) as usize;
        let count = u32_at(16) as usize;

        let stride = 8 + taps * 4 * 2;
        if BLOB.len() != 20 + count * stride {
            return Err(format!(
                "hrir blob is {} bytes, expected {}",
                BLOB.len(),
                20 + count * stride
            ));
        }

        let mut responses = Vec::with_capacity(count);
        let mut at = 20;
        for _ in 0..count {
            let azimuth = f32_at(at);
            let elevation = f32_at(at + 4);
            let mut p = at + 8;
            let read = |p: &mut usize| {
                let mut v = Vec::with_capacity(taps);
                for _ in 0..taps {
                    v.push(f32_at(*p));
                    *p += 4;
                }
                v
            };
            let left = read(&mut p);
            let right = read(&mut p);
            responses.push(Response { azimuth, elevation, left, right });
            at += stride;
        }
        Ok(Set { rate, taps, responses })
    }

    // A non-zero result means a layout gained an angle the blob lacks.
    pub fn nearest(&self, azimuth: f32, elevation: f32) -> (&Response, f32) {
        let target = unit(azimuth, elevation);
        let mut best = &self.responses[0];
        let mut best_dot = -2.0f32;
        for r in &self.responses {
            let d = dot(target, unit(r.azimuth, r.elevation));
            if d > best_dot {
                best = r;
                best_dot = d;
            }
        }
        (best, best_dot.clamp(-1.0, 1.0).acos().to_degrees())
    }
}

fn unit(az: f32, el: f32) -> [f32; 3] {
    let (a, e) = (az.to_radians(), el.to_radians());
    [e.cos() * a.cos(), e.cos() * a.sin(), e.sin()]
}

fn dot(a: [f32; 3], b: [f32; 3]) -> f32 {
    a[0] * b[0] + a[1] * b[1] + a[2] * b[2]
}
